ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/Impl/ChainSel.hs

```diff
@@ -709,16 +709,26 @@ chainSelectionForBlock cdb@CDB{..} blockCache hdr punish = electric $ do
         curTip      = castPoint $ AF.headPoint curChain
         curHead     = AF.headAnchor curChain
 
-    -- Either frag extends loe or loe extends frag
-    --
-    -- PRECONDITION: @AF.withinFragmentBounds (AF.anchorPoint frag) loe@
+    -- Either frag extends loe or loe extends frag. The fragment is represented
+    -- by the current chain and a diff with that current chain. It is tempting
+    -- to only consider the suffix of the diff, but that would be incorrect,
+    -- because the diff might not intersect with the LoE fragment, because the
+    -- diff suffix is anchored somewhere on the current chain and LoE frag's tip
+    -- might be older than that anchor.
     followsLoEFrag :: LoE (AnchoredFragment (Header blk))
-                   -> AnchoredFragment (Header blk)
-                   -> Bool
-    followsLoEFrag LoEDisabled _ = True
-    followsLoEFrag (LoEEnabled loe) frag =
-         AF.withinFragmentBounds (AF.headPoint loe) frag
-      || AF.withinFragmentBounds (AF.headPoint frag) loe
+                   -> ChainAndLedger blk
+                   -> ChainDiff (Header blk)
```

Comment:
True; it used to trim some candidates and reject others, but now it just trims everything. Removing the `Maybe` monad means we introduce the precondition that the LoE intersects with the current selection, but that is enforced in other parts of the code, so that sounds good to me. I will write a version that returns directly a `ChainDiff (Header blk)`.

Reply:
7b023f1dfed2c4f5c76dbcc252a63792817aa152